### PR TITLE
Better UDP and TCP listener to be resilient to socket crash

### PR DESCRIPTION
**Pull request by ajacquemot**
### What does this PR do?

Improved errors recovery for the listeners:
- Restart a TCP listener when an error occured
- Restart a UDP connection if it crashed because of a too big message

### Motivation

Ensure the agent can keep forwarding logs even if errors network errors occurred.

### Additional Notes

Unit-test are missing

### CODE REVIEW

File: pkg/logs/input/listener/tcp.go [hunk 1]
```diff
@@ -32,9 +31,8 @@ func NewTCPListener(pp pipeline.Provider, source *config.LogSource) (*TCPListene
 		return nil, err
 	}
 	source.Status.Success()
-	connHandler := NewConnectionHandler(pp, source)
+	connHandler := NewConnectionHandler(pp, source, func() (net.Conn, error) { return nil, fmt.Errorf("dropping connection") })
```
Comment (tmichelet) [inline pkg/logs/input/listener/tcp.go]: could we add a small comment on why we use this function that returns an error?

**Review by mnshdw — CHANGES_REQUESTED** [review]
Good change but I have a few concerns

File: pkg/logs/input/listener/tcp.go (hunk 1 above)
Comment (ajacquemot) [reply pkg/logs/input/listener/tcp.go]: Agreed, I changed the logic though

File: pkg/logs/input/listener/worker.go [hunk 2]
```diff
@@ -82,10 +84,10 @@ func (w *Worker) readForever() {
 			// stop reading data from the connection
 			return
 		default:
-			if !w.mustKeepConnAlive() {
+			if !w.keepAlive {
 				w.conn.SetReadDeadline(time.Now().Add(defaultTimeout))
 			}
-			inBuf := make([]byte, 4096)
+			inBuf := make([]byte, 1024)
```
[the remark this answers is not in the document]
Comment (ajacquemot) [reply pkg/logs/input/listener/worker.go]: This is an error, I reverted the change.

File: pkg/logs/input/listener/worker.go [hunk 3]
```diff
@@ -22,24 +22,27 @@ const defaultTimeout = 60000 * time.Millisecond
 
 // Worker reads data from a connection
```
[the remark this answers is not in the document]
Comment (ajacquemot) [reply pkg/logs/input/listener/worker.go]: I think we should rename it as `Tailer` for future refactoring.

**Review by mnshdw — APPROVED** [review]

**Pull request merged**